File: _scripts/voice-rag-etl/etl.py
#!/usr/bin/env python3
"""
ETL Script — Nandini (案内) Voice RAG

Zero-dependency Python ETL (uses only stdlib):
  - Reads knowledge.txt
  - Chunks text with overlap
  - Embeds via Google AI REST API (text-embedding-004)
  - Upserts to Pinecone REST API

Usage:
  1. Create a .env file with GOOGLE_API_KEY, PINECONE_API_KEY, PINECONE_INDEX
  2. python3 etl.py
"""

import json
import os
import re
import sys
import time
import urllib.request
import urllib.error
from pathlib import Path

# ─── Config ─────────────────────────────────────────────────────
EMBEDDING_MODEL = "gemini-embedding-001"  # supports outputDimensionality
EMBEDDING_DIM = 768
CHUNK_SIZE = 500
CHUNK_OVERLAP = 50
UPSERT_BATCH = 5
SCRIPT_DIR = Path(__file__).parent


# ─── Load .env ──────────────────────────────────────────────────
def load_env():
    env_file = SCRIPT_DIR / ".env"
    if env_file.exists():
        for line in env_file.read_text().splitlines():
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if "=" not in line:
                continue
            key, val = line.split("=", 1)
            os.environ.setdefault(key.strip(), val.strip())


load_env()

GOOGLE_API_KEY = os.environ.get("GOOGLE_API_KEY", "")
PINECONE_API_KEY = os.environ.get("PINECONE_API_KEY", "")
PINECONE_INDEX = os.environ.get("PINECONE_INDEX", "")

if not all([GOOGLE_API_KEY, PINECONE_API_KEY, PINECONE_INDEX]):
    print("\n❌  Missing env vars. Required: GOOGLE_API_KEY, PINECONE_API_KEY, PINECONE_INDEX")
    print("   Create a .env file in this directory.\n")
    sys.exit(1)


# ─── HTTP Helper ────────────────────────────────────────────────
def api_post(url, data, headers=None):
    body = json.dumps(data).encode("utf-8")
    req = urllib.request.Request(url, data=body, method="POST")
    req.add_header("Content-Type", "application/json")
    if headers:
        for k, v in headers.items():
            req.add_header(k, v)
    try:
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        err_body = e.read().decode("utf-8") if e.fp else ""
        raise RuntimeError(f"HTTP {e.code}: {err_body}") from e


def api_get(url, headers=None):
    req = urllib.request.Request(url)
    if headers:
        for k, v in headers.items():
            req.add_header(k, v)
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode("utf-8"))


# ─── Embedding with retry ───────────────────────────────────────
def embed_text(text, max_retries=5):
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{EMBEDDING_MODEL}:embedContent?key={GOOGLE_API_KEY}"
    data = {
        "model": f"models/{EMBEDDING_MODEL}",
        "content": {"parts": [{"text": text}]},
        "outputDimensionality": EMBEDDING_DIM,
    }
    for attempt in range(max_retries):
        try:
            result = api_post(url, data)
            return result["embedding"]["values"]
        except RuntimeError as e:
            if "429" in str(e) and attempt < max_retries - 1:
                wait = 2 ** (attempt + 1)  # 2, 4, 8, 16, 32 seconds
                print(f"\n     ⏳ Rate limited. Waiting {wait}s...", flush=True)
                time.sleep(wait)
            else:
                raise


# ─── Pinecone ───────────────────────────────────────────────────
def get_pinecone_host():
    url = f"https://api.pinecone.io/indexes/{PINECONE_INDEX}"
    data = api_get(url, headers={"Api-Key": PINECONE_API_KEY})
    return data["host"]


def pinecone_upsert(host, vectors):
    url = f"https://{host}/vectors/upsert"
    return api_post(url, {"vectors": vectors}, headers={"Api-Key": PINECONE_API_KEY})


# ─── Chunker ────────────────────────────────────────────────────
def chunk_text(text):
    chunks = []
    start = 0
    while start < len(text):
        end = min(start + CHUNK_SIZE, len(text))
        if end < len(text):
            slc = text[start:end]
            breaks = [slc.rfind(". "), slc.rfind(".\n"), slc.rfind("\n\n")]
            best = max(breaks)
            if best > CHUNK_SIZE * 0.3:
                end = start + best + 1
        chunks.append({"text": text[start:end].strip(), "offset": start})
        # Ensure forward progress (prevent infinite loop near end of text)
        new_start = end - CHUNK_OVERLAP
        if new_start <= start:
            new_start = end  # Skip overlap for the last chunk
        start = new_start
        if start >= len(text):
            break
    return [c for c in chunks if c["text"]]


# ─── Metadata ───────────────────────────────────────────────────
def extract_sections(text):
    sections = []
    current = {"project_name": "General", "github_url": ""}
    section_start = 0
    offset = 0
    for line in text.split("\n"):
        pm = re.match(r"^Project(?:/Research)?:\s*(.+)$", line, re.I)
        if pm:
            if offset > section_start:
                sections.append({"start": section_start, "end": offset, "meta": dict(current)})
            current = {"project_name": pm.group(1).strip(), "github_url": ""}
            section_start = offset
        gm = re.search(r"GitHub Link:\s*(https?://[^\s*]+)", line, re.I)
        if gm:
            current["github_url"] = gm.group(1).rstrip("*")
        offset += len(line) + 1
    sections.append({"start": section_start, "end": offset, "meta": dict(current)})
    return sections


def meta_for(sections, offset):
    for s in sections:
        if s["start"] <= offset < s["end"]:
            return s["meta"]
    return {"project_name": "General", "github_url": ""}


# ─── Main ───────────────────────────────────────────────────────
def main():
    print("\n🚀  Nandini ETL Pipeline — Starting...\n")

    # 1. Load
    kpath = SCRIPT_DIR / "knowledge.txt"
    raw = kpath.read_text(encoding="utf-8")
    print(f"📄  Loaded knowledge.txt ({len(raw)} chars)")

    # 2. Sections
    sections = extract_sections(raw)
    print(f"📑  {len(sections)} sections:")
    for s in sections:
        print(f"     → {s['meta']['project_name']}", flush=True)

    # 3. Chunk
    chunks = chunk_text(raw)
    print(f"\n✂️   {len(chunks)} chunks (size={CHUNK_SIZE}, overlap={CHUNK_OVERLAP})")

    # 4. Embed
    print(f"\n🧠  Embedding via {EMBEDDING_MODEL}...")
    embeddings = []
    for i, c in enumerate(chunks):
        vals = embed_text(c["text"])
        embeddings.append(vals)
        print(f"     {i+1}/{len(chunks)} done", end="\r", flush=True)
        if i < len(chunks) - 1:
            time.sleep(1.5)  # Stay under free-tier rate limit
    print(f"\n     ✓ {len(embeddings)} embeddings (dim={len(embeddings[0])})")

    # 5. Pinecone
    print(f"\n🌲  Connecting to Pinecone: {PINECONE_INDEX}")
    host = get_pinecone_host()
    print(f"     Host: {host}")

    # 6. Build vectors
    vectors = []
    for i, c in enumerate(chunks):
        m = meta_for(sections, c["offset"])
        vectors.append({
            "id": f"annai-{i}",
            "values": embeddings[i],
            "metadata": {
                "text": c["text"],
                "project_name": m["project_name"],
                "github_url": m["github_url"],
            },
        })

    # 7. Upsert
    print(f"\n⬆️   Upserting {len(vectors)} vectors...")
    t0 = time.time()
    for i in range(0, len(vectors), UPSERT_BATCH):
        batch = vectors[i : i + UPSERT_BATCH]
        pinecone_upsert(host, batch)
        done = min(i + UPSERT_BATCH, len(vectors))
        print(f"     {done}/{len(vectors)}", end="\r")

    elapsed = time.time() - t0
    print(f"\n\n✅  Done! {len(vectors)} vectors upserted in {elapsed:.1f}s")
    print("🎉  Nandini's knowledge base is ready!\n")


if __name__ == "__main__":
    main()

File: _scripts/voice-rag-etl/etl.js
/**
 * ETL Script — Maya (案内) Voice RAG
 * 
 * ZERO external dependencies — uses only built-in fetch().
 * Works on Node.js 18+ (including v25).
 * 
 *   - Reads knowledge.txt
 *   - Chunks text with overlap
 *   - Embeds via Google AI REST API (text-embedding-004)
 *   - Upserts to Pinecone REST API
 * 
 * Usage:
 *   1. Create a .env file with GOOGLE_API_KEY, PINECONE_API_KEY, PINECONE_INDEX
 *   2. node etl.js
 */

import { readFileSync, existsSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

// ─── Load .env manually (no dotenv dependency) ─────────────────
const __dirname = dirname(fileURLToPath(import.meta.url));
const envPath = resolve(__dirname, '.env');
if (existsSync(envPath)) {
  const envContent = readFileSync(envPath, 'utf-8');
  for (const line of envContent.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const eqIndex = trimmed.indexOf('=');
    if (eqIndex === -1) continue;
    const key = trimmed.substring(0, eqIndex).trim();
    let val = trimmed.substring(eqIndex + 1).trim();
    if (val.startsWith('"') && val.endsWith('"')) {
      val = val.substring(1, val.length - 1);
    } else if (val.startsWith("'") && val.endsWith("'")) {
      val = val.substring(1, val.length - 1);
    }
    if (!process.env[key]) process.env[key] = val;
  }
}

// ─── Configuration ──────────────────────────────────────────────
const EMBEDDING_MODEL = 'gemini-embedding-001'; // 768 dimensions
const CHUNK_SIZE = 800;
const CHUNK_OVERLAP = 100;
const UPSERT_BATCH = 5;

const GOOGLE_API_KEY = process.env.GOOGLE_API_KEY;
const PINECONE_API_KEY = process.env.PINECONE_API_KEY;
const PINECONE_INDEX = process.env.PINECONE_INDEX;

// ─── Validate ──────────────────────────────────────────────────
if (!GOOGLE_API_KEY || !PINECONE_API_KEY || !PINECONE_INDEX) {
  console.error('\n❌  Missing env vars. Required: GOOGLE_API_KEY, PINECONE_API_KEY, PINECONE_INDEX');
  console.error('   Create a .env file in this directory.\n');
  process.exit(1);
}

// ─── Google Embedding (raw fetch) ──────────────────────────────
async function embedText(text) {
  const url = `https://generativelanguage.googleapis.com/v1beta/models/${EMBEDDING_MODEL}:embedContent?key=${GOOGLE_API_KEY}`;
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: `models/${EMBEDDING_MODEL}`,
      content: { parts: [{ text }] },
      outputDimensionality: 768,
    }),
  });
  if (!res.ok) {
    throw new Error(`Embedding API ${res.status}: ${await res.text()}`);
  }
  const data = await res.json();
  return data.embedding.values;
}

// ─── Pinecone REST API ─────────────────────────────────────────
async function getPineconeHost() {
  const res = await fetch(`https://api.pinecone.io/indexes/${PINECONE_INDEX}`, {
    headers: { 'Api-Key': PINECONE_API_KEY },
  });
  if (!res.ok) {
    throw new Error(`Pinecone index lookup failed ${res.status}: ${await res.text()}`);
  }
  const data = await res.json();
  return data.host;
}

async function pineconeUpsert(host, vectors) {
  const res = await fetch(`https://${host}/vectors/upsert`, {
    method: 'POST',
    headers: {
      'Api-Key': PINECONE_API_KEY,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ vectors }),
  });
  if (!res.ok) {
    throw new Error(`Pinecone upsert failed ${res.status}: ${await res.text()}`);
  }
  return res.json();
}

async function pineconeDeleteAll(host) {
  const res = await fetch(`https://${host}/vectors/delete`, {
    method: 'POST',
    headers: {
      'Api-Key': PINECONE_API_KEY,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ deleteAll: true }),
  });
  if (!res.ok) {
    throw new Error(`Pinecone delete failed ${res.status}: ${await res.text()}`);
  }
  return res.json();
}

// ─── Text Chunker ──────────────────────────────────────────────
function chunkText(text) {
  const chunks = [];
  let start = 0;
  while (start < text.length) {
    let end = Math.min(start + CHUNK_SIZE, text.length);
    if (end < text.length) {
      const slice = text.substring(start, end);
      const lastBreak = Math.max(
        slice.lastIndexOf('. '),
        slice.lastIndexOf('.\n'),
        slice.lastIndexOf('\n\n')
      );
      if (lastBreak > CHUNK_SIZE * 0.3) end = start + lastBreak + 1;
    }
    chunks.push({ text: text.substring(start, end).trim(), offset: start });
    if (end >= text.length) break;
    start = end - CHUNK_OVERLAP;
  }
  return chunks.filter((c) => c.text.length > 0);
}

// ─── Metadata Extraction ───────────────────────────────────────
function extractSections(text) {
  const sections = [];
  const lines = text.split('\n');
  let current = { project_name: 'General', github_url: '' };
  let sectionStart = 0;
  let offset = 0;

  for (const line of lines) {
    const pm = line.match(/^Project(?:\/Research)?:\s*(.+)$/i);
    if (pm) {
      if (offset > sectionStart) {
        sections.push({ start: sectionStart, end: offset, meta: { ...current } });
      }
      current = { project_name: pm[1].trim(), github_url: '' };
      sectionStart = offset;
    }
    const gm = line.match(/GitHub Link:\s*(https?:\/\/[^\s*]+)/i);
    if (gm) current.github_url = gm[1].replace(/\*$/, '');
    offset += line.length + 1;
  }
  sections.push({ start: sectionStart, end: offset, meta: { ...current } });
  return sections;
}

function metaFor(sections, offset) {
  for (const s of sections) {
    if (offset >= s.start && offset < s.end) return s.meta;
  }
  return { project_name: 'General', github_url: '' };
}

// ─── Main ──────────────────────────────────────────────────────
async function main() {
  console.log('\n🚀  Maya ETL Pipeline — Starting...\n');

  // 1. Load knowledge
  const kPath = resolve(__dirname, 'knowledge.txt');
  const rawText = readFileSync(kPath, 'utf-8');
  console.log(`📄  Loaded knowledge.txt (${rawText.length} chars)`);

  // 2. Sections
  const sections = extractSections(rawText);
  console.log(`📑  ${sections.length} sections:`);
  sections.forEach((s) => console.log(`     → ${s.meta.project_name}`));

  // 3. Chunk
  const chunks = chunkText(rawText);
  console.log(`\n✂️   ${chunks.length} chunks (size=${CHUNK_SIZE}, overlap=${CHUNK_OVERLAP})`);

  // 4. Embed
  console.log(`\n🧠  Embedding via ${EMBEDDING_MODEL}...`);
  const embeddings = [];
  for (let i = 0; i < chunks.length; i++) {
    const values = await embedText(chunks[i].text);
    embeddings.push(values);
    process.stdout.write(`     ${i + 1}/${chunks.length} done\r`);
    if (i < chunks.length - 1) await sleep(250);
  }
  console.log(`\n     ✓ ${embeddings.length} embeddings (dim=${embeddings[0].length})`);

  // 5. Pinecone host
  console.log(`\n🌲  Connecting to Pinecone index: ${PINECONE_INDEX}`);
  const host = await getPineconeHost();
  console.log(`     Host: ${host}`);

  // 6. Build vectors
  const vectors = chunks.map((c, i) => {
    const m = metaFor(sections, c.offset);
    return {
      id: `annai-${i}`,
      values: embeddings[i],
      metadata: { text: c.text, project_name: m.project_name, github_url: m.github_url },
    };
  });

  // 7. Upsert
  console.log(`\n🧹  Wiping all previous vectors from Pinecone...`);
  await pineconeDeleteAll(host);
  console.log(`     ✓ Index cleared.`);

  console.log(`\n⬆️   Upserting ${vectors.length} new vectors...`);
  const t0 = Date.now();
  for (let i = 0; i < vectors.length; i += UPSERT_BATCH) {
    const batch = vectors.slice(i, i + UPSERT_BATCH);
    await pineconeUpsert(host, batch);
    process.stdout.write(`     ${Math.min(i + UPSERT_BATCH, vectors.length)}/${vectors.length}\r`);
  }
  const secs = ((Date.now() - t0) / 1000).toFixed(1);
  console.log(`\n\n✅  Done! ${vectors.length} vectors upserted in ${secs}s`);
  console.log(`🎉  Maya's knowledge base is ready!\n`);
}

function sleep(ms) { return new Promise((r) => setTimeout(r, ms)); }

main().catch((e) => { console.error('\n❌  Failed:', e.message); process.exit(1); });
